src/ui/backend/canvas/FilterRenderer.js

```diff
@@ -0,0 +1,158 @@
+		unusedCanvas = removedEntry ? removedEntry.value : null;
+
+		return resultImg;
+	};
+
+
+	this.renderColorFilter = function (ctx, srcImg, srcX, srcY, srcW, srcH, color, op) {
+		var result = this.getCanvas(srcW, srcH);
+		var resultCtx = result.getContext('2d');
+		// render the base image
+		resultCtx.globalCompositeOperation = 'source-over';
+		srcImg.render(resultCtx, srcX, srcY, srcW, srcH, 0, 0, srcW, srcH);
+		// render the filter color
+		resultCtx.globalCompositeOperation = op;
+		resultCtx.fillStyle = "rgba(" + color.r  + "," + color.g + "," + color.b + "," + color.a + ")";
+		resultCtx.fillRect(0, 0, srcW, srcH);
+		// use our base image to cut out the image shape from the rect
+		resultCtx.globalCompositeOperation = 'destination-in';
+		srcImg.render(resultCtx, srcX, srcY, srcW, srcH, 0, 0, srcW, srcH);
+		return result;
+	};
+
+	this.renderMultiply = function(ctx, srcImg, srcX, srcY, srcW, srcH, filter) {
+		var color = filter.get();
+		var result = this.getCanvas(srcW, srcH);
+		var resultCtx = result.getContext('2d');
+		var imgData = srcImg.getImageData(srcX, srcY, srcW, srcH);
+		var data = imgData.data;
+		// simplified multiply math outside of the massive for loop
+		var a = color.a;
+		var mr = 1 + a * ((color.r / 255) - 1);
+		var mg = 1 + a * ((color.g / 255) - 1);
+		var mb = 1 + a * ((color.b / 255) - 1);
+		for (var i = 0, len = data.length; i < len; i += 4) {
+		  data[i] *= mr;
+		  data[i + 1] *= mg;
+		  data[i + 2] *= mb;
+		}
+		// put the updated rgb data into our new canvas
+		resultCtx.putImageData(imgData, 0, 0);
+		return result;
+	};
+
+	this.renderMask = function (ctx, srcImg, srcX, srcY, srcW, srcH, mask, op) {
+		var result = this.getCanvas(srcW, srcH);
+		var resultCtx = result.getContext('2d');
+		// render the mask image
+		var srcMaskX = mask.getSourceX();
+		var srcMaskY = mask.getSourceY();
+		var srcMaskW = mask.getSourceW();
+		var srcMaskH = mask.getSourceH();
+		resultCtx.globalCompositeOperation = 'source-over';
+		mask.render(resultCtx, srcMaskX, srcMaskY, srcMaskW, srcMaskH, 0, 0, srcW, srcH);
+		// render the base image
+		resultCtx.globalCompositeOperation = op;
+		srcImg.render(resultCtx, srcX, srcY, srcW, srcH, 0, 0, srcW, srcH);
+		return result;
+	};
+
+	this.clearCache = function () {
+		this.cache.removeAll();
+	};
+
+	this.getCacheKey = function(url, srcX, srcY, srcW, srcH, filter) {
+		var filterType = filter.getType();
+		var suffix;
+		if (filterType === 'NegativeMask' || filterType === 'PositiveMask') {
+			suffix = filter.getMask().getURL();
+		} else {
+			var color = filter.get();
+			var alpha = (color.a * 255) & 0xff;
+			suffix = "" + alpha + "|" + color.r + "|" + color.g + "|" + color.b;
+		}
+		var cacheKey = filterType + "|" + url + "|" + srcX + "|" + srcY + "|" + srcW + "|" + srcH + "|" + suffix;
+		return cacheKey;
+	};
+
+	this.getCanvas = function(width, height) {
+		var result = unusedCanvas || new Canvas();
```

Comment:
Let's pass in `{ width: width, height: height }` to the `Canvas` constructor and call `.resize(width, height)` on `unusedCanvas` if it exists.